ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp

```diff
@@ -0,0 +1,1210 @@
+                    const AxisVector order = get_default_order(input_shape);
+
+                    ngraph::runtime::reference::reshape(
+                        reinterpret_cast<const char*>(input->get_data_ptr<T>()),
+                        reinterpret_cast<char*>(output->get_data_ptr<T>()),
+                        input_shape,
+                        order,
+                        new_shape,
+                        element_type.size());
+                    return output;
+                }
+
+                template <typename T>
+                void contract_two_inputs(HostTensorVector& inputs,
+                                         std::vector<std::string>& input_subscripts,
+                                         const std::string& output_subscript,
+                                         size_t input_ind1,
+                                         size_t input_ind2)
+                {
+                    // assume that input_ind1 < input_ind2 without loss of generality, otherwise,
+                    // just swap them
+                    if (input_ind2 < input_ind1)
+                    {
+                        std::swap(input_ind1, input_ind2);
+                    }
+
+                    // perform sanity check for arguments
+                    auto num_inputs = inputs.size();
+                    NGRAPH_CHECK(num_inputs == input_subscripts.size(),
+                                 "Each input must have own subscript.");
+                    NGRAPH_CHECK(input_ind2 < num_inputs && input_ind1 != input_ind2,
+                                 "Incorrect input index is specified.");
+
+                    const auto& input1 = inputs[input_ind1];
+                    const auto& input2 = inputs[input_ind2];
+
+                    // extract diagonals in case repeated labels in the corresponding input
+                    // subscripts
+                    extract_diagonal<T>(inputs, input_subscripts, input_ind1);
+                    extract_diagonal<T>(inputs, input_subscripts, input_ind2);
+
+                    // reduce dimensions for input operands if possible
+                    reduce_input<T>(inputs, input_subscripts, output_subscript, input_ind1);
+                    reduce_input<T>(inputs, input_subscripts, output_subscript, input_ind2);
+
+                    // step 0. split dimensions of both operands into three groups:
+                    // 1. dimension indices with the same labels (in both subscripts) that are NOT
+                    // reduced - common labels (dimensions)
+                    // 2. dimension indices with labels that are met only in one of two subscripts -
+                    // separate labels (dimensions)
+                    // 3. dimension indices with the same labels (in both subscripts) that are
+                    // reduced - reduced labels (dimensions) NOTE: dimension is reduced iff. the
```

Comment:
`dimension is reduced iff` a typo?

Reply:
iff = if and only if:). But anyway, I will re-write it to `if`

Reply:
Fixed.